packages/nimble-components/src/table/tests/table-matrix.stories.ts

```diff
@@ -63,10 +67,15 @@ export const tableThemeMatrix: Story = createMatrixThemeStory(
     createMatrix(component)
 );
 
-tableThemeMatrix.play = (): void => {
-    document.querySelectorAll<Table>('nimble-table').forEach(table => {
+tableThemeMatrix.play = async (): Promise<void> => {
+    const tables = document.querySelectorAll<Table>('nimble-table');
+    tables.forEach(table => {
         table.setData(data);
     });
+    await waitForUpdatesAsync();
```

Comment:
hmhmhmmhmhmhmm this feels wrong. For method calls like this I definitely expect to be able to call them synchronously to each other.

Users definitely shouldn't need waitForUpdatesAsync calls in there, that is effectively just a random wait. Maybe whatever is requiring it would take longer in the future. Just spells intermittency.

Even making them async, ie.:
```js
table.setData(data);
await table.setSelectedRecordIds(['1']);
```

Is a bit of a bummer. With an async task it could be hard to reason about the state. Maybe something else runs or a user click finishes that modifies the selected results.

It's not super clear to me from source inspection where the async operation is being inserted.

Reply:
The issue I ran into here was this code path:
- `connectedCallback()` calls `this.updateTracker.trackAllStateChanged()`
- as part of FAST's async queue, the update tracker calls `update()` on the table
- `update()` resets the selection state because one of the states it sets is the selection mode, and we change the selection mode, we also reset the selection

In summary, the need for the wait here is to have the element be connected when `setSelectedRecordIds()` is called. But, I agree, this isn't a great a solution to that problem. I'll brainstorm some more on a way to allow table clients to be able to call `setSelectedRecordIds()` before connection without having that state be cleared when the table becomes connected.

Reply:
One approach to resolving this issue is to have `setData()` and `setSelectedRecordIds()` go through the `UpdateTracker`. I put together a branch that shows what this would look like: https://github.com/ni/nimble/commit/a2758ff4b4dc9ba8edc862f8abb1c4cac7b6fc22

Some pros and cons of switching to this approach are:
**Pros**
- Further unifies the code paths that update TanStack
- Avoids updating TanStack multiple times when `setData()` or `setSelectedRecordIds()` is called prior to connection
- Fixes potential timing issues with `setSelectedRecordIds()` called before the `id-field-name` has been applied

**Cons**
- Makes the `update()` code more complex. Moreover, it isn't clear to me yet what the code will look like to avoid updating the selection state when row IDs are invalid (note: this isn't handled in the branch I referenced above).
- Calling `setSelectedRecordIds()` doesn't apply immediately, so a user has to wait before calling `getSelectedRecordIds()`
    - Note: we could potentially make `setSelectedRecordIds()` `async` to help hide this from the user, but I'm not sure if that is a good idea because the returned Promise may not resolve for some time if the table element isn't connect
- Similar to the above con, calling `setData()` doesn't immediately apply and update the selection, so the user has to wait before calling `getSelectedRecordIds()` to see the impact the new data had on their selection


Another alternative would be for `setData()` and `setSelectedRecordIds()` to not normally (or maybe ever?) go through the `UpdateTracker`. But, when `setData()` and `setSelectedRecordIds()` are called prior to connection, those two states are cached rather than immediately applied, and then those cached values are applied as part of the first `update()` call after connection.

I think I lean towards the second alternative (the one without a branch associated with it) because it solves the problem at hand without introducing timing issues when `setData()` and `setSelectedRecordIds()` are called after connection. But, I'm interested in other opinions -- @rajsite @atmgrifter00

Reply:
Option one is tough because we lose the synchronous validation behaviors. It feels like it maybe takes us back to the issue of these being declarative apis. The biggest concern there being how you make the getter behave reasonably. Maybe in that scenario only the getSelectedRecordIds() is async?

I'm a little concerned for alternative 2 how complex this offline cache will get. It'll handle setData, setSelectedRecordIds, getSelectedRecordIds? What if the user toggles the selection mode, ie. setData -> selectionMode = multiple -> setSelectedRecordIds ->  selectionMode = single -> getSelectedRecordIds, will it resolve the changed selected rows state? Will we need to cache expanded rows eventually? Maybe the getters don't work until after connected? Do they throw before connected?

What if setData, getSelectedRecordIds, and setSelectedRecordIds were all async. They could do something like:
1. throw if not connected to the DOM
2. await a `DOM.nextUpdate()` to give confidence pending bindings and dom operations are done (existing property sets on table element or child column elements are done)
3. Maybe check that the updatetracker does not having pending work at this point? if so go back to 2?
4. After the queue has settled, check the table validity and throw if invalid
5. finally set the data / set/get-SelectedRowIds

Reply:
That suggestion does simplify the table's code, but I think it will also make our clients' jobs harder. They will have to be more aware of the lifecycle of the table such that they don't call one of those functions before the table is connected. I'm not sure that we have a way for them to check that, do we?

Reply:
What if for step 1 we don't throw and just wait ourselves?

If someone explicitly does the imperative:

```ts
const table  = document.createElement('nimble-table');
await table.setData();
```

it will hang and they should call append first, but the majority use-case is like here, i.e. some templating engine is rendering the nodes asynchronously and giving us a reference that will eventually be attached?

I guess an alternative path to go down is that HTML elements are expected to behave well before they are even attached. ie. I can do:

```ts
const img = document.createElement('img');
img.addEventListener('load', () => console.log('loaded'));
img.src= 'someurl';
```

and without attaching to the dom it should behave well.

So maybe another question is how far off are we from being able to do that? I think a big part is that templates are not rendered yet, maybe we can do in setData, etc:

1. If `this.$fastcontroller.isConnected === false` then run `$fastcontroller.onConnected()`;
2. do the rest of the stuff above (DOM.queue update, check update tracker, check validity, etc.)
3. finally do the thing

I think once onConnected is called to construct the template then all the mutation observers, etc in the tempate, ie the `children` directive etc should come to life:
![image](https://user-images.githubusercontent.com/1588923/229861517-d7483bfc-981c-452e-b340-1126b692bd79.png)

Web components cannot construct new HTML etc in the Constructor (limitation by design) so most do it the first time it is connected but that's not a requirement. Nothing in [their onConnected implementation](https://github.com/microsoft/fast/blob/1be91f1b6212a55963cdaaa0cd8af76b4d7091fa/packages/web-components/fast-element/src/components/controller.ts#L279) concerns me about calling it early.

In the table maybe the virtualizer would continue to be constructed in the connectedCallback but the state tracking and columnObservation could maybe be split out to run sooner.

Reply:
Along those lines, we could queue updates for state on the table from the table's constructor, and then we could queue updates for state coming from columns from the `connectedCallback` where we start observing column changes. The async `setData`, `setRecordIds`, and `getRecordIds` would just need to wait until the update queue was empty and then perform their tasks.

Reply:
Glad that worked out 🎉 some of these other options were sounding pretty complex for us / the users 😅